src/codegen/codegen-expression.lisp

```diff
@@ -179,9 +179,68 @@
     (declare (type tc:environment env))
     `(return-from ,(continue-label (node-continue-label expr))))
 
+  (:method ((node node-catch) env)
+    (declare (type tc:environment env))
+    (let* ((block-label   (gensym "CATCH-BLOCK"))
+           (handler-cases
+             (loop
+               :for branch :in (node-catch-branches node)
+               :for pattern
+                 := (catch-branch-pattern branch)
+               ;; wildcard and variable patterns catch all 'error cases
+               :for exception-name
+                 := (etypecase pattern
+                      (pattern-constructor 
+                       (let* ((ctor-name              (pattern-constructor-name pattern))
+                              (ctor                   (tc::lookup-constructor env ctor-name)))
+                         (tc:constructor-entry-classname ctor)))
+                      ((or pattern-wildcard pattern-var)
+                       'error))
+               :for case-body
+                 := (codegen-expression (catch-branch-body branch) env)
+               :for lambda-var
+                 := (gensym (symbol-name exception-name))
+               :for bindings
+                 := (nth-value 1 (codegen-pattern pattern lambda-var (pattern-type pattern) env))
+               ;; NB: if CASE-BODY invokes a restart then control will
+               ;; be transferred before the transfer due to
+               ;; return-from.
+               :for inner-body
+                 := `(return-from ,block-label ,case-body)
+               :collect `(,exception-name
+                          (lambda (,lambda-var)
+                            (declare (ignorable ,lambda-var))
+                            (let ,bindings
+                              (declare (ignorable ,@(mapcar #'car bindings)))
+                              ,inner-body))))))
+      `(block ,block-label
+         (handler-bind ,handler-cases
+           ,(codegen-expression (node-catch-expr node) env)))))
+
+  (:method ((node node-resumable) env)
+    (declare (type tc:environment env))
+    (let* ((clauses
+             (loop
+               :for branch :in (node-resumable-branches node)
+               :for pattern
+                 := (resumable-branch-pattern branch)
+               :for restart-name := (tc:lisp-type (pattern-type pattern) env)
+               :for restart-var := (gensym (symbol-name restart-name))
+               :for bindings := (nth-value 1 (codegen-pattern pattern restart-var (pattern-type pattern) env))
+               :for inner-body := (codegen-expression (resumable-branch-body branch) env)
+
+               :when bindings
```

Comment:
I meant *this* won't work. It won't do what I want. It will miss cases where constrctors are matched with all wildcard or literal patterns.